execution/gethexec/espresso_finality_node.go

```diff
@@ -0,0 +1,158 @@
+*/
+type EspressoFinalityNode struct {
+	stopwaiter.StopWaiter
+
+	config       SequencerConfigFetcher
+	execEngine   *ExecutionEngine
+	hotshotState *HotsShotState
+	namespace    uint64
+}
+
+func NewEspressoFinalityNode(execEngine *ExecutionEngine, configFetcher SequencerConfigFetcher) *EspressoFinalityNode {
+	config := configFetcher()
+	if err := config.Validate(); err != nil {
+		panic(err)
+	}
+	return &EspressoFinalityNode{
+		execEngine:   execEngine,
+		config:       configFetcher,
+		namespace:    config.EspressoFinalityNodeConfig.Namespace,
+		hotshotState: NewHotsShotState(config.EspressoFinalityNodeConfig.HotShotUrl, config.EspressoFinalityNodeConfig.StartBlock),
+	}
+}
+
+func (n *EspressoFinalityNode) createBlock(ctx context.Context) (returnValue bool) {
+
+	if n.hotshotState.nextSeqBlockBum == 0 {
+		latestBlock, err := n.hotshotState.client.FetchLatestBlockHeight(ctx)
+		if err != nil && latestBlock == 0 {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return false
+		}
+		log.Info("Started espresso finality node at the latest hotshot block", "block number", latestBlock)
+		n.hotshotState.nextSeqBlockBum = latestBlock
+	}
+
+	nextSeqBlockNum := n.hotshotState.nextSeqBlockBum
+	header, err := n.hotshotState.client.FetchHeaderByHeight(ctx, nextSeqBlockNum)
+	if err != nil {
+		arbos.LogFailedToFetchHeader(nextSeqBlockNum)
+		return false
+	}
+
+	arbTxns, err := n.hotshotState.client.FetchTransactionsInBlock(ctx, header.Height, n.namespace)
+	if err != nil {
+		arbos.LogFailedToFetchTransactions(header.Height, err)
+		return false
+	}
+
+	arbHeader := &arbostypes.L1IncomingMessageHeader{
+		Kind:        arbostypes.L1MessageType_L2Message,
+		Poster:      l1pricing.BatchPosterAddress,
+		BlockNumber: header.L1Head,
+		Timestamp:   header.Timestamp,
+		RequestId:   nil,
+		L1BaseFee:   nil,
+	}
+
+	// Deserialize the transactions and ignore the malformed transactions
+	txes := types.Transactions{}
+	for _, tx := range arbTxns.Transactions {
+		var out types.Transaction
+		if err := out.UnmarshalBinary(tx); err != nil {
+			log.Warn("Malformed tx is found")
+			continue
+		}
+		txes = append(txes, &out)
+	}
+
+	hooks := arbos.NoopSequencingHooks()
+	_, err = n.execEngine.SequenceTransactions(arbHeader, txes, hooks, false)
+	if err != nil {
+		log.Error("Espresso Finality Node: Failed to sequence transactions", "err", err)
+		return false
+	}
+	n.hotshotState.advance()
+
+	return true
+}
+
+func (n *EspressoFinalityNode) Start(ctx context.Context) error {
+	n.StopWaiter.Start(ctx, n)
+	err := n.CallIterativelySafe(func(ctx context.Context) time.Duration {
+		madeBlock := n.createBlock(ctx)
+		if madeBlock {
+			return 0
+		}
+		return retryTime
+	})
+	if err != nil {
+		return fmt.Errorf("failed to start espresso finality node: %w", err)
+	}
+	return nil
+}
+
+func (n *EspressoFinalityNode) PublishTransaction(ctx context.Context, tx *types.Transaction, options *arbitrum_types.ConditionalOptions) error {
```

Comment:
Finiality node should just forward the transactions to sovereign sequencer or do nothing

Reply:
I am not sure I understand the flow correctly then, if we send the messages through the sovereign sequencer - it will be classified as a `Espresso Message`. In that case, the batch poster will detect the same and send the transaction to espresso again for finality (code [here](https://github.com/EspressoSystems/nitro-espresso-integration/blob/integration/arbnode/batch_poster.go#L1321))

Reply:
Current behaviour is similar to what we had [before](https://github.com/EspressoSystems/nitro-espresso-integration/blob/ef2c5ccd2cd700eed461d7fe039ee471ca3090b9/execution/gethexec/espresso_sequencer.go#L129)

Reply:
Currently `sovereign sequencer` also needs to be developed more to adapt to the latest design. On the one hand, the sovereign sequencer will receive the user transactions and then send them to the marketplace builder. On the other hand, sovereign sequencer (or its batch poster) is also responsible for building the correct batches to the base layer. The correctness of these batches is guaranteed by the sequencer inbox contract.
For the finality node, the only thing it needs to do is to fetch the HotShot blocks and build its blocks. Since these transactions are from HotShot, not from the base layer, it can provide fast finality.
BTW, with the `marketplace builder`, I think the `/submit/submit` endpoint which used in the `go client` becomes useless. Previously we sequenced all the user transactions by receiving them permisionlessly, and now it is not a case.

Reply:
Thank you, I think I understand the flow now. 

After thinking about it more, I believe we should not implement `PublishTransaction` in `espresso finality node` and my reasoning is as follows:

1. Espresso finality node will be used by bridges etc to listen to transactions that have been finalized by hotshot and will in my belief not be used to `publish` transactions
2. If we do want to implement `publish` transactions in our `espresso finality node`, the flow will be as follows `espresso finality node` will be creating blocks with `hotshot` finalized transactions and will also be `creating blocks` using the `sovereign sequencer` because `PublishTransaction` function will forward the transactions to the `sovereign sequencer`. It will then be hard for the client to understand which `blocks` are the espresso finalized blocks and which arent. So it will be better for the `Espresso Finality node` to not publish transactions

Let me know your thoughts!

Reply:
CC @ImJeremyHe @sveitser @zacshowa - feel free to comment your thoughts on this

Reply:
I agree that it doesn't make sense to do something that causes the same blocks to be produced multiple times. One question I have is if we are forwarding blocks produced by this node to the sovereign sequencer, won't they be sent to the espresso network again? Naively this seems to be the case, and doesn't feel like correct behavior. 

If the above is true, I don't think we need to be re-sending these blocks through the sequencer before we alter it to handle this new type of message so we aren't caught in a loop of re-sending transactions. Instead, if I'm not off base, I think it would suffice to potentially store these blocks in some lightweight KV store, and then just provide the ability to search for them over an RPC as mentioned in #235 

I'm not sure what reason we have to want to for forwarding transactions to the nitro sequencer from this node as mentioned in the original ticket, but perhaps I just need context.

Additionally, if we do end up forwarding things to the sequencer with this node, we should make sure that we can gate any behavior that forwards transactions to the espresso network behind the chain config :smile:

Reply:
This function is not forwarding blocks, but forwarding transactions. This function doesn't create any blocks, it just receives and forwards. So there will not be any duplication transactions.